README.md
![GitHub Actions tests](https://github.com/johnjohndoe/fdroid-build-checker/workflows/Test/badge.svg) [![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black) [![Imports: isort](https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=ef8336)](https://pycqa.github.io/isort/) [![Checked with mypy](http://www.mypy-lang.org/static/mypy_badge.svg)](http://mypy-lang.org/) [![AGPL 3.0](https://img.shields.io/badge/license-AGPL%203.0-blueviolet.svg)](https://www.gnu.org/licenses/agpl-3.0.html)



# fdroid-build-checker

Python code to check if [F-Droid][fdroid] recent build has succeeded or failed.

![Logo](gfx/logo.png)

## Description

The scripts check if your custom packages occur on the [recent changes][recent-changes] page and then inspect the
associated last build pages for the build result.

## Setup

```sh
# Install dependencies
pipenv install --dev

# Setup pre-commit and pre-push hooks
pipenv run pre-commit install -t pre-commit
pipenv run pre-commit install -t pre-push
```

## Install

```sh
$ pip3 install --user git+https://github.com/johnjohndoe/fdroid-build-checker.git
```


## Usage

```sh
$ python3 ~/.local/lib/python3.8/site-packages/fdroid_build_checker/
```

See [`fdroid_build_checker/__main__.py`](fdroid_build_checker/__main__.py)

## Tests

Execute the following command to run all tests:

```sh
pipenv run python -m pytest
```

## Dependencies

See [Pipfile](Pipfile)


## Related projects

- *BuildNotify*, a CCMenu/CCTray equivalent for Ubuntu ([Source code][build-notify-sources], [Website][build-notify-website])
- *F-Droid Build Status* app for Android ([Source code][fdroid-build-status-sources], [F-Droid][fdroid-build-status-fdroid])


## License

```
Copyright (C) 2021  Tobias Preuss

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
```

## Credits

This package was created with Cookiecutter and
the [sourcery-ai/python-best-practices-cookiecutter](https://github.com/sourcery-ai/python-best-practices-cookiecutter)
project template.


[build-notify-sources]: https://github.com/anaynayak/buildnotify
[build-notify-website]: https://anaynayak.github.io/buildnotify/
[fdroid]: https://f-droid.org
[fdroid-build-status-fdroid]: https://f-droid.org/en/packages/de.storchp.fdroidbuildstatus/
[fdroid-build-status-sources]: https://codeberg.org/pstorch/F-Droid_Build_Status
[recent-changes]: https://f-droid.org/wiki/index.php?title=Special:RecentChanges&days=30&from=&hidebots=0&hideanons=1&hideliu=1&limit=500


CHANGELOG.md
# CHANGELOG

## 2021-10-12

- Bump lxml from 4.6.2 to 4.6.3.

## 2021-06-11

- Prevent crash cause by a connection timeout.

## 2021-01-23

- Configure package to be installable.
- Initial version.


test/test_fdroid_build_checker.py
from fdroid_build_checker.fdroid_build_checker import fib


def test_fib() -> None:
    assert fib(0) == 0
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(3) == 2
    assert fib(4) == 3
    assert fib(5) == 5
    assert fib(10) == 55


test/parsers/test_last_build_html_parser.py
from unittest import TestCase

from fdroid_build_checker.parsers.last_build_html_parser import LastBuildHtmlParser


class TestLastBuildHtmlParser(TestCase):
    def test_parse_success(self):
        parser = LastBuildHtmlParser(
            url="https://f-droid.org/wiki/page/info.metadude.android.fosdem.schedule/lastbuild_81"
        )
        last_build_result = parser.parse()
        self.assertTrue(last_build_result.is_success())

    # def test_parse_failure(self):
    #     parser = LastBuildHtmlParser(
    #         url="https://f-droid.org/wiki/page/info.metadude.android.rc3.schedule/lastbuild_81"
    #     )
    #     last_build_result = parser.parse()
    #     self.assertFalse(last_build_result.is_success())


test/parsers/test_recent_changes_html_parser.py
from unittest import TestCase

from fdroid_build_checker.parsers.recent_changes_html_parser import (
    RecentChangesHtmlParser,
    RecentChange,
)


class TestRecentChangesHtmlParser(TestCase):
    def test_parse(self):
        package_names = [
            "info.metadude",
        ]
        parser = RecentChangesHtmlParser(
            domain_url="https://f-droid.org",
            url="https://f-droid.org/wiki/index.php?title=Special:RecentChanges&days=30&from=&hidebots=0&hideanons=1"
            "&hideliu=1&limit=500",
        )
        recent_changes = parser.parse(package_names)
        if len(recent_changes) > 0:
            for recent_change in recent_changes:
                self.assert_recent_change(recent_change)

    def assert_recent_change(self, recent_change: RecentChange):
        self.assertTrue(recent_change.date != "")
        self.assertTrue(recent_change.time != "")
        self.assertTrue(recent_change.package != "")
        self.assertTrue(recent_change.version_code != "")
        self.assertTrue(recent_change.href != "")


fdroid_build_checker/__main__.py
from typing import List

from fdroid_build_checker.parsers.last_build_html_parser import (
    LastBuild,
    LastBuildHtmlParser,
)
from fdroid_build_checker.parsers.recent_changes_html_parser import (
    RecentChange,
    RecentChangesHtmlParser,
)


def parse_recent_change(package_names: List[str]) -> List[RecentChange]:
    parser = RecentChangesHtmlParser(
        domain_url="https://f-droid.org",
        url="https://f-droid.org/wiki/index.php?title=Special:RecentChanges&days=30&from=&hidebots=0&hideanons=1"
        "&hideliu=1&limit=500",
    )
    return parser.parse(package_names)


def print_result(recent_change: RecentChange, last_build: LastBuild) -> None:
    print(f"{recent_change}\n{last_build}")


def main():
    package_names = [
        "info.metadude",
    ]
    recent_changes = parse_recent_change(package_names)
    for recent_change in recent_changes:
        if not recent_change.is_redirect():
            last_build = LastBuildHtmlParser(recent_change.href).parse()
            print_result(recent_change, last_build)


if __name__ == "__main__":
    main()


fdroid_build_checker/parsers/last_build_html_parser.py
import sys
import urllib.error
import urllib.request

from lxml import etree


class LastBuild:
    url: str = ""
    parsed_text: str = ""

    def __init__(self, url: str, parsed_text: str):
        self.url = url
        self.parsed_text = parsed_text

    def is_success(self) -> bool:
        return self.parsed_text == "Build succeeded"

    def __str__(self) -> str:
        return f"{self.is_success()}, '{self.parsed_text}', {self.url}"


class LastBuildHtmlParser:
    """Parser specialized for reading from the F-Droid last build page."""

    def __init__(self, url: str):
        """
        Initializes the parser with the given URL.

        :param url: The URL of the HTML page.
        """
        self.url = url
        try:
            schedule_data = urllib.request.urlopen(url)
            parser = etree.HTMLParser(recover=True)
            tree = etree.parse(schedule_data, parser)
        except (ValueError, urllib.error.HTTPError, urllib.error.URLError) as e:
            print(f"Error ({e}) loading file from URL: {url}")
            sys.exit(1)
        self.root = tree.getroot()

    def parse(self) -> LastBuild:
        """
        Parses the last page to retrieve its last build result.
        Returns a LastBuild or throws an exception.
        """
        for build_log_headline in self.root.iter("h2"):
            for child in build_log_headline.getchildren():
                if child.tag == "span" and child.get("id") == "Build_Log":
                    paragraph = build_log_headline.getnext()
                    return LastBuild(self.url, paragraph.text.strip())
        raise Exception(f"Failure parsing HTML from url = '{self.url}'.")


def main():
    parser = LastBuildHtmlParser(
        url="https://f-droid.org/wiki/page/info.metadude.android.fosdem.schedule/lastbuild_81"
    )
    print(parser.parse())

    parser = LastBuildHtmlParser(
        url="https://f-droid.org/wiki/page/info.metadude.android.rc3.schedule/lastbuild_81"
    )
    print(parser.parse())


if __name__ == "__main__":
    main()


fdroid_build_checker/parsers/recent_changes_html_parser.py
import sys
import urllib.error
import urllib.request
from typing import Dict, List

from lxml import etree


class RecentChange:
    date: str = ""
    time: str = ""
    package: str = ""
    version_code: str = "None"
    href: str = ""

    def __init__(
        self, date: str, time: str, package: str, version_code: str, href: str
    ):
        self.date = date
        self.time = time
        self.package = package
        if version_code is not None and version_code != "":
            self.version_code = version_code
        self.href = href

    def is_redirect(self) -> bool:
        return self.version_code == "None"

    def __str__(self) -> str:
        return f"{self.date}, {self.time}, {self.package}, {self.version_code}, {self.href}"


def destructure_link_text(link_text: str) -> Dict[str, str]:
    """
    Returns a dictionary containing a package and a version code value.
    """
    components = dict()
    parts = link_text.split("/lastbuild")
    components["package"] = parts[0].strip()
    if len(parts) == 2:
        components["version_code"] = parts[1].strip()
    else:
        components["version_code"] = ""
    return components


class RecentChangesHtmlParser:
    """Parser specialized for reading from the F-Droid recent changes page."""

    def __init__(self, domain_url: str, url: str):
        """
        Initializes the parser with the given URL.

        :param domain_url: The domain part with protocol of the url.
        :param url: The URL of the HTML page.
        """
        self.domain_url = domain_url
        try:
            schedule_data = urllib.request.urlopen(url)
            parser = etree.HTMLParser(recover=True)
            tree = etree.parse(schedule_data, parser)
        except (ValueError, urllib.error.HTTPError, urllib.error.URLError) as e:
            print(f"Error ({e}) loading file from URL: {url}")
            sys.exit(1)
        self.root = tree.getroot()

    def parse(self, package_names: List[str]) -> List[RecentChange]:
        """
        Parses the rows for the given package names and returns a list of RecentChange.
        The returned list might be empty if no rows are present on the page.
        """
        recent_changes = []

        for changes_list_div in self.root.iter("div"):
            if changes_list_div.get("class") == "mw-changeslist":
                date = ""
                for child in changes_list_div.getchildren():
                    if child.tag == "h4":
                        date = child.text.strip()
                    if child.tag == "div":

                        for table in child.iter("table"):
                            table_classes = table.get("class")
                            if (
                                "mw-enhanced-rc" in table_classes
                                and "mw-changeslist-line-not-watched" in table_classes
                            ):

                                time_stamp = ""
                                for tr in table.iter("tr"):
                                    # print(etree.tostring(tr, method="html", pretty_print=True))

                                    for td in tr.iter("td"):
                                        if td.get("class") == "mw-enhanced-rc":
                                            # print(etree.tostring(td, method="html", pretty_print=True))
                                            for abbr in td.iter("abbr"):
                                                time_stamp = abbr.tail.strip()

                                        for span in td.iter("span"):
                                            # print(etree.tostring(span, method="html", pretty_print=True))
                                            if span.get("class") == "mw-title":

                                                for a in span.iter("a"):
                                                    href = a.get("href")
                                                    # print(etree.tostring(a, method="html", pretty_print=True))
                                                    if (
                                                        a.get("class")
                                                        == "mw-changeslist-title"
                                                    ):
                                                        for (
                                                            package_name
                                                        ) in package_names:
                                                            link_text = a.text.strip()
                                                            if (
                                                                package_name
                                                                in link_text
                                                            ):
                                                                (
                                                                    package,
                                                                    version_code,
                                                                ) = destructure_link_text(
                                                                    link_text
                                                                ).values()
                                                                recent_change = RecentChange(
                                                                    date=date,
                                                                    time=time_stamp,
                                                                    package=package,
                                                                    version_code=version_code,
                                                                    href=f"{self.domain_url}{href}",
                                                                )
                                                                recent_changes.append(
                                                                    recent_change
                                                                )
        return recent_changes


def main():
    package_names = [
        "info.metadude",
    ]
    parser = RecentChangesHtmlParser(
        domain_url="https://f-droid.org",
        url="https://f-droid.org/wiki/index.php?title=Special:RecentChanges&days=30&from=&hidebots=0&hideanons=1"
        "&hideliu=1&limit=500",
    )
    recent_changes = parser.parse(package_names)
    print(*recent_changes, sep="\n")


if __name__ == "__main__":
    main()
